Add load balancer and IP pool docs (#323)
harvester/docs: docs/rancher/cloud-provider.md

  ![](/img/v1.2/rancher/rke2-cloud-provider.png)

### Deploying to the RKE2 Custom Cluster

1. Use `generate_addon.sh` to generate cloud config and place it into directory `/var/lib/rancher/rke2/etc/config-files/cloud-config` on every node.

   ```
   curl -sfL https://raw.githubusercontent.com/harvester/cloud-provider-harvester/master/deploy/generate_addon.sh | bash -s <serviceaccount name> <namespace> 
   ```

  :::note

  The `generate_addon.sh` script depends on `kubectl` and `jq` to operate the Harvester cluster

  The script needs access to the `Harvester Cluster` kubeconfig to work.

  The namespace needs to be the namespace in which the guest cluster will be created.

  :::

> 2. Select the `Harvester` cloud provider

Reviewer: ```suggestion
2. Select the `Harvester` cloud provider.
```

  

### IPAM

> Harvester's built-in load balancer supports both `DHCP` and `Pool` mode. You can select the mode in the Rancher UI. Harvester adds the annotation `cloudprovider.harvesterhci.io/ipam` to the service behind. Additionally, Harvester cloud provider provides a special mode `Share IP` where a service will share its load balancer IP with other services.

Reviewer: ```suggestion
Harvester's built-in load balancer supports both **DHCP** and **Pool** modes and you can select the mode in the Rancher UI. Harvester adds the annotation `cloudprovider.harvesterhci.io/ipam` to the service. Additionally, Harvester cloud provider provides a special **Share IP** mode where a service will share its load balancer IP with other services.
```

> Harvester's built-in load balancer supports both **DHCP** and **Pool** modes and you can select the mode in the Rancher UI. Harvester adds the annotation `cloudprovider.harvesterhci.io/ipam` to the service. Additionally, Harvester cloud provider provides a special **Share IP** mode where a service will share its load balancer IP with other services.

Reviewer: ```suggestion
Harvester's built-in load balancer supports both **DHCP** and **Pool** modes, and you can select the mode in the Rancher UI. Harvester adds the annotation `cloudprovider.harvesterhci.io/ipam` to the service. Additionally, Harvester cloud provider provides a special **Share IP** mode where a service will share its load balancer IP with other services.
```

> - DCHP:  A DHCP server is required. The Harvester Load Balancer controller will request an IP address from the DHCP server.

Reviewer: ```suggestion
- **DCHP: ** A DHCP server is required. The Harvester load balancer controller will request an IP address from the DHCP server.
```

> - Pool: You need an IP pool configured in the Harvester UI. The Harvester Load Balancer controller will allocate an IP for the load balancer service following [the IP pool selection policy](/networking/ippool.md/#selection-policy).

Reviewer: ```suggestion
- **Pool:** You need an IP pool configured in the Harvester UI. The Harvester load balancer controller will allocate an IP for the load balancer service following [the IP pool selection policy](/networking/ippool.md/#selection-policy).
```

> - Share IP: When creating a new load balancer service, you can select an existing load balancer service to get its load balancer IP. The new service is called secondary service, and the selected service is called primary service. You can specify the primary service in the secondary service by annotation `cloudprovider.harvesterhci.io/primary-service`. There are two limitations.

Reviewer: ```suggestion
- Share IP: When creating a new load balancer service, you can select an existing load balancer service to get its load balancer IP. The new service is called the secondary service, and the existing chosen service is called the primary service. You can specify the primary service in the secondary service by annotation `cloudprovider.harvesterhci.io/primary-service`. There are two limitations:
```

> - Share IP: When creating a new load balancer service, you can select an existing load balancer service to get its load balancer IP. The new service is called the secondary service, and the existing chosen service is called the primary service. You can specify the primary service in the secondary service by annotation `cloudprovider.harvesterhci.io/primary-service`. There are two limitations:

Reviewer: ```suggestion
- **Share IP:** When creating a new load balancer service, you can select an existing load balancer service to get its load balancer IP. The new service is called the secondary service, and the currently chosen service is called the primary service. You can specify the primary service in the secondary service by annotation `cloudprovider.harvesterhci.io/primary-service`. There are two limitations:
```

  - It's not allowed for the secondary service to share the IP with other services.

>   - All the services sharing the same IP could not have duplicated ports.

Reviewer: ```suggestion
  - All the services sharing the same IP can not have duplicated ports.
```

> - **Share IP:** When creating a new load balancer service, you can select an existing load balancer service to get its load balancer IP. The new service is called the secondary service, and the currently chosen service is called the primary service. You can specify the primary service in the secondary service by annotation `cloudprovider.harvesterhci.io/primary-service`. There are two limitations:
>   - It's not allowed for the secondary service to share the IP with other services.
>   - All the services sharing the same IP can not have duplicated ports.

Reviewer 2: ```suggestion
- **Share IP:** When creating a new load balancer service, you can re-utilize an existing load balancer service IP. The new service is called a secondary service, while the presently chosen service is considered the primary service. To specify the primary service in the secondary service, you can add the annotation `cloudprovider.harvesterhci.io/primary-service: $primary-service-name`.  However, there are two known limitations
  - Secondary services cannot share their IP with additional services.
  - Services sharing the same IP address must not possess duplicated ports.
```

- **Pool:** You need an IP pool configured in the Harvester UI. The Harvester load balancer controller will allocate an IP for the load balancer service following [the IP pool selection policy](/networking/ippool.md/#selection-policy).

- **Share IP:** When creating a new load balancer service, you can select an existing load balancer service to get its load balancer IP. The new service is called the secondary service, and the currently chosen service is called the primary service. You can specify the primary service in the secondary service by annotation `cloudprovider.harvesterhci.io/primary-service`. There are two limitations:
  - It's not allowed for the secondary service to share the IP with other services.
  - All the services sharing the same IP can not have duplicated ports.

:::note

It is not allowed to modify the IPAM mode. You need to create a new service if you want to modify the IPAM mode.

:::

> ### Health Checks

Reviewer 2: prefer to add a note that `Health Checks` is not supported in v1.2.0 instead of just removing it directly.

Author: I add a note.

Author: Add more explanation about Health Checks.
